[codec] Allow negative field numbers in structs/unions

jesboat opened the pull request:
Summary:

Negative field IDs are strongly discouraged, but some old services used
them and are stuck with them to maintain wire compatibility with older
clients. Don't error out on them.

Test Plan:

grep in swift-codec for `> 0`, `>= 0`, `< 0`, `<= 0`, (and the other way
around). A bunch of ThriftStructs which didn't used to be able to be
loaded are now.

Reviewers: alandau, andrewcox, haijunz

CC:

jesboat committed:
[codec] Allow negative field numbers in structs/unions

Summary:

Negative field IDs are strongly discouraged, but some old services used
them and are stuck with them to maintain wire compatibility with older
clients. Don't error out on them. Add an `isLegacyId` attribute to the
`@ThriftField` annotation to allow them when necessary. Any
`@ThriftField` which has a negative ID must have `isLegacyId=true`;
`@ThriftField`'s with non-negative IDs must not (but may specify
`isLegacyId=false` if they want.)

Test Plan:

grep in swift-codec for `> 0`, `>= 0`, `< 0`, `<= 0`, (and the other way
around). A bunch of ThriftStructs which didn't used to be able to be
loaded are now. Added unit tests.

Reviewers: alandau, andrewcox, haijunz

CC:

dain commented:
Can you update the existing tests or add new tests for this?

andrewcox commented:
To encourage proper struct design going forward, can we make this only work if you have specified a backward-compatibility override in the @ThriftField annotation to allow this?

dain commented:
@andrewcox maybe instead make it a config option that is set on the ThriftCatalog or ThriftCodecManager when they are constructed.  That way you have to turn on support for legacy applications.

andrewcox commented:
@dain I was thinking about that too, but I like annotating individual @ThriftFields (or @ThriftStructs) because someone using that class doesn't need to think about it, and at the same time, if we have some legacy interface that uses a negative field id that a lot of services end up needing to call, having the override be per class/field instead of per-app will protect people working on the app from adding new negative ids without realizing it is wrong.

dain commented:
Seems reasonable

electrum commented:
Agreed, adding an option like `@ThriftField(value = -5, legacyId = true)` seems like the way to go. And only allow the `legacyId = true` if the ID is actually negative.

jesboat commented:
Updated to require `isLegacyId=true` for negative field IDs. The rules are: any `@ThriftField` annotation with a negative ID needs `isLegacyId=true`; any `@ThriftField` annotation with a non-negative ID needs `isLegacyId` left out (defaults to false) or `isLegacyId=false`; any `@ThriftField` annotation not specifying an ID should have `isLegacyId` left out. The rules are intended to enforce that any thrift field has isLegacyId=true iff it has a negative ID and that isLegacyId=true is specified syntacticly on the same `@ThriftField` annotation which sets the ID.

Added test cases.

jesboat commented:
> I guess I'm ok with the name 'isLegacyId' for the field, but use the word 'has a negative field' (not has a legacy field) in the error message to be more clear.

Originally I went with @electrum's suggestion of `legacyFieldId` in the annotation, which reads slightly more fluently when using `@ThriftField`s. The downside was referring to it during the metadata processing. Call it `legacyId` and then `getLegacyId()` is a confusing getter. Call it `isLegacyId` and then the name changes between the annotation and the metadata processing code, which increases confusion. I ended up deciding that it needed to be something consistent, and `isLegacyId` seemed to work slightly better overall than `legacyId`.

Changed the error messages as you suggest.

> better comment than "ick"? :)

k

> Code looks fine, except that I don't see the need to treat this as optional, since it has a default value (and we can't full distinguish between the set and not set case for legacyId==false anyway).

Basically, we're in the same situation as why `getThriftFieldId()` wants to return an `Optional<Short>`. Consider something like:

    @ThriftField(id=-4, isLegacyId=true)
    public boolean getBlah() { ... }
    
    @ThriftField
    public void setBlah(boolean v) { ... }

If we always returned the value of isLegacyId, then above would fail because one of the annotations has isLegacyId=true and one has false. So you'd need to write it as:

    @ThriftField(id=-4, isLegacyId=true)
    pubic boolean getBlah() { ...}
    
    @ThriftField(isLegacyId=true)
    public void setBlah(boolean v) { ... }

But now we have an annotation which has `isLegacyId=true` without a field ID, which is weird, and also isLegacyId now behaves different from id. So you'd need to write

    @ThriftField(id=-4, isLegacyId=true)
    pubic boolean getBlah() { ...}
    
    @ThriftField(id=-4, isLegacyId=true)
    public void setBlah(boolean v) { ... }

The end result being that, if a thrift field had a negative ID, both the id and isLegacyId=true would need to be specified in *every* applicable `@ThriftField` annotation, which would kinda suck.
